test/db_test.py

```diff
@@ -0,0 +1,358 @@
+		def datesPopulated():
+			db.addUser(test_user)
+			row = db.DB_CONN.execute('''
+				SELECT created_at, updated_at, accessed_at
+				FROM Users WHERE name = ?
+			''', [test_name]).fetchone()
+
+			assertDateNearNow(row[0])
+			assertDateNearNow(row[1])
+			#assertDateNearNow(row[2])
+
+		@it('Cannot have multiple users with the same name')
+		def duplicateName():
+			db.addUser(test_user)
+			assert_that(
+				calling(db.addUser).with_args(test_user),
+				raises(IntegrityError, 'UNIQUE constraint failed: Users.name')
+			)
+
+		@it('Password hash and salt required')
+		def hashAndSaltRequired():
+			assert_that(
+				calling(db.addUser).with_args({
+					'name': test_name,
+					'pass_hash': test_hash,
+					'pass_salt': None
+				}),
+				raises(IntegrityError, 'NOT NULL constraint failed: Users.pass_salt')
+			)
+			assert_that(
+				calling(db.addUser).with_args({
+					'name': test_name,
+					'pass_hash': None,
+					'pass_salt': test_salt
+				}),
+				raises(IntegrityError, 'NOT NULL constraint failed: Users.pass_hash')
+			)
+
+	@describe('Update User')
+	def updateUser():
+
+		added_user = None
+
+		@beforeEach
+		def _beforeEach():
+			nonlocal added_user
+			row = addTestUser(whole_row=True)
+			added_user = {
+				'id': row[0],
+				'name': row[1],
+				'pass_hash': row[2],
+				'pass_salt': row[3],
+				'email': row[4],
+				'created_at': row[5],
+				'updated_at': row[6],
+				'accessed_at': row[7]
+			}
+
+		@afterEach
+		def _afterEach():
+			cleanup()
+
+		@it('Update user email preserves other fields')
+		def updatedUser():
+			nonlocal added_user
+			update_email = 'new@email.com'
+			added_user['email'] = update_email
+			res = db.updateUser(added_user)
+			updated_user = db.getUser(test_name)
+
+			assert_that(updated_user.get('name'), equal_to(test_name))
+			assert_that(updated_user.get('pass_hash'), equal_to(test_hash))
+			assert_that(updated_user.get('pass_salt'), equal_to(test_salt))
+			assert_that(updated_user.get('email'), equal_to(update_email))
+
+		@it('Date modified changed on update')
+		def modifiedUpdated():
+			sleep(1) # Delay to ensure modified time is different
+			db.updateUser(test_user)
+			update_user = db.getUser(test_name)
+			assert_that(
+				update_user.get('updated_at'),
+				not_(equal_to(test_user.get('updated_at')))
+			)
+
+		@it('Date created not changed on update')
+		def createdNotUpdated():
+			sleep(1) # Same here
+			db.updateUser(test_user)
+			update_user = db.getUser(test_name)
+			assert_that(update_user.get('created_at'), test_user.get('created_at'))
+
+	@describe('Add Code')
+	def addCode():
+
+		@beforeEach
+		def _beforeEach():
+			nonlocal test_id
+			test_id = addTestUser()
+
+		@afterEach
+		def _afterEach():
+			cleanup()
+
+		@it('None not allowed for code')
+		def codeNone():
+			nonlocal test_id
+			assert_that(
+				calling(db.addEmailCode).with_args(None, test_id, test_email),
+				raises(IntegrityError, 'NOT NULL constraint failed: Codes.code')
+			)
```

Comment:
Would it be worth it to also make sure empty strings cannot be passed? Ex: "". In my eyes, this would essentially be the same as NULL, but I don't know if the database agrees with that.